cmd/clusterctl/pkg/client/repository/repository_github.go

```diff
@@ -0,0 +1,301 @@
+
+const (
+	httpsScheme              = "https"
+	githubDomain             = "github.com"
+	githubTokeVariable       = "github-token"
+	githubReleaseRepository  = "releases"
+	githubLatestReleaseLabel = "latest"
+)
+
+// gitHubRepository provides support for providers hosted on GitHub.
+//
+// We support GitHub repositories that use the release feature to publish artifacts and versions.
+// Repositories must use versioned releases, including the "latest" meta version
+// (https://help.github.com/en/github/administering-a-repository/linking-to-releases#linking-to-the-latest-release).
+type gitHubRepository struct {
+	providerConfig           config.Provider
+	configVariablesClient    config.VariablesClient
+	authenticatingHTTPClient *http.Client
+	owner                    string
+	repository               string
+	defaultVersion           string
+	rootPath                 string
+	componentsPath           string
+	injectClient             *github.Client
+}
+
+var _ Repository = &gitHubRepository{}
+
+// DefaultVersion returns defaultVersion field of gitHubRepository struct
+func (g *gitHubRepository) DefaultVersion() string {
+	return g.defaultVersion
+}
+
+// RootPath returns rootPath field of gitHubRepository struct
+func (g *gitHubRepository) RootPath() string {
+	return g.rootPath
+}
+
+// ComponentsPath returns componentsPath field of gitHubRepository struct
+func (g *gitHubRepository) ComponentsPath() string {
+	return g.componentsPath
+}
+
+// GetFile returns a file for a given provider version
+func (g *gitHubRepository) GetFile(version, path string) ([]byte, error) {
+	release, err := g.getReleaseByTag(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get GitHub release %s", version)
+	}
+
+	// download files from the release
+	files, err := g.downloadFilesFromRelease(release, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download files from GitHub release %s", version)
+	}
+
+	return files, nil
+}
+
+// newGitHubRepository returns a gitHubRepository implementation
+func newGitHubRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*gitHubRepository, error) {
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	// Check if the url is a github repository
+	if !(rURL.Scheme == httpsScheme && rURL.Host == githubDomain) {
+		return nil, errors.New("invalid url: a GitHub repository url should start with https://github.com")
+	}
+
+	// Check if the repository Path is in the expected format.
+	t := strings.Split(strings.TrimPrefix(rURL.Path, "/"), "/")
+	if len(t) < 5 || t[2] != githubReleaseRepository {
+		return nil, errors.Errorf(
+			"invalid url: a GitHub repository url should be in the form https://github.com/{owner}/{Path}/%s/{latest|version-tag}/{componentsClient.yaml}",
```

Comment:
Ok to rename Path into Project or Repository (I'm +1 on Repository)
I'm -1 to the second change because {path/components.yaml} does not reflect the correct order

Reply:
Addressed